### PR TITLE
ISPN-6122 AbstractCacheTransaction is prone to race conditions: NPE i…

### PR DESCRIPTION
**Pull request by rhusar**
…n containsLockOrBackupLock(Object)

Jira
https://issues.jboss.org/browse/ISPN-6122

### CODE REVIEW

**Comment by pruivo** [comment]
looks good. testing and pushing...

**Comment by pruivo** [comment]
integrated! thanks @rhusar !

**Comment by rhusar** [comment]
Thanks!

File: core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java [hunk 1]
```diff
@@ -211,33 +212,34 @@ public int getTopologyId() {
 
    @Override
    public void addBackupLockForKey(Object key) {
-      // we need to synchronize this collection to be able to get a valid snapshot from another thread during state transfer
-      if (backupKeyLocks == null) backupKeyLocks = Collections.synchronizedSet(new HashSet<>(INITIAL_LOCK_CAPACITY));
-      backupKeyLocks.add(key);
+      // we need a synchronized collection to be able to get a valid snapshot from another thread during state transfer
+      final Set<Object> keys = backupKeyLocks.updateAndGet((value) -> value == null ? Collections.synchronizedSet(new HashSet<>(INITIAL_LOCK_CAPACITY)) : value);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java]: No need for synchronization between writers, we already have a `TransactionSynchronizerInterceptor` dealing with that.

File: core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java [hunk 2]
```diff
@@ -211,33 +212,34 @@ public int getTopologyId() {
...
       if (trace) log.tracef("Registering locked key: %s", toStr(key));
-      lockedKeys.add(key);
+      keys.add(key);
    }
 
    @Override
    public Set<Object> getLockedKeys() {
-      return lockedKeys == null ? InfinispanCollections.emptySet() : lockedKeys;
+      final Set<Object> keys = lockedKeys.get();
+      return keys == null ? InfinispanCollections.emptySet() : keys;
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/transaction/impl/AbstractCacheTransaction.java]: I believe making a copy of the `lockedKeys` reference in a local variable would have been enough.

**Pull request closed without merging**